### PR TITLE
docs: Add documentation for rp2 DMA support.

### PR DESCRIPTION
**Pull request by nickovs**
This PR provides documentation for the rp2.DMA support in PR #7641 and the associated additions to rp2.PIO in #7650.

### CODE REVIEW

**Review by Maffels — APPROVED** [review]

**Comment by nickovs** [comment]
@Maffels Thanks for approving this. Does this mean that #7641 and #7650 might get approved soon? Updating the documentation is of little use without incorporating the code changes being documented! 😀

**Comment by Maffels** [comment]
My apologies, @nickovs , I tagged "approve" by accident and am unaware of any way to rectify this mistake. 
As far as contributing to this project: I'm afraid I'm just enthusiastic about this (together with #7641 and #7650) being added to the next micropython release but not related to any of this in any other way. The "approved" status I accidentally gave it is thus worthless.
Feel free to remove this comment and perhaps if possible the "approved" status.

**Comment by nickovs** [comment]
I have pushed an update to reflect the changes to the configuration model that @dpgeorge requested for #7641.

**Comment by NewWheelTech** [comment]
The documentation says "# The count is in transfers, which default to four-byte words, so divide length by 4". 

I cannot find this anywhere in the Pico documentation, testing the Pico it appears to default to 0, which would be bytes.

**Comment by mendenm** [comment]
@nickovs A picky note:  you might do a global search/substitute of 'crtl' to 'ctrl'.  In quite a few places, the letters are swapped.

**Comment by nickovs** [comment]
@NewWheelTech The [code as merged](https://github.com/micropython/micropython/blob/9feb0689eeaca5ce88aedcc680f997a3b4d0221c/ports/rp2/rp2_dma.c#L304) defaults to word transfers, so the docs are correct for what we've got. I guess there might be an argument for changing the code to match the C SDK.

**Comment by dpgeorge** [comment]
> I guess there might be an argument for changing the code to match the C SDK.

It's not the SDK that defaults to byte transfers, but rather the register is reset to 0 and that means byte transfer size.

I think it's sensible for the default CTRL value (DEFAULT_DMA_CONFIG) to default to word-sized transfers, because that's the most common use (I guess, eg for peripheral access).  Then maybe the `count` parameter to `.config()` could always be measured in bytes to make it easier for the user... although that would be harder to implement because if the user sets the CTRL register after specifying `count` then the count value needs to change based on the new data size set by CTRL.

**Comment by mendenm** [comment]
I would recommend not making the ```count``` parameter be bytes.  That  would make the hardware documentation invalid, and I suspect most of us learn about the DMA functionality from that.  Also, other DMA devices I have worked with (e.g. on dsPIC33 processors) count transfers, not bytes.

**Comment by nickovs** [comment]
The problem with making `count` be in bytes is that it's neoither what the hardware uses nor what the C SDK uses. While we can fix this for the `.config` method and for reading and writing the `count` attribute, we can't fix this for direct access to the register bank, so if people are doing any sort of scatter/gather operation then it will be inconsistent. Also, if people are reading examples in the the RP2040 datasheet then they will have to make semantic modifications rather than just transliterate. As a result I really think that it should stay as transfers rather than bytes.

**Comment by dpgeorge** [comment]
> As a result I really think that it should stay as transfers rather than bytes.

Yes, agreed.

**Comment by dpgeorge** [comment]
@nickovs let me know if/when this is ready for final review and merge.

**Comment by nickovs** [comment]
With #7650 merged I think that this now accurately reflects the code, so please go ahead with a review and merge.

**Comment by pi-mst** [comment]
Thanks for your efforts @nickovs, much appreciated! 

However, is it possible to come up with a more tangible example? It'd be great to be able to point folks at, say, a [Wokwi](https://wokwi.com/)-hosted example that provided a concise, clear and compelling use case - the examples in the docs either execute practically immediately or are quite low-level and difficult to appreciate. Maybe something like a PIO/DMA solution to control a WS2812B...

**Comment by mendenm** [comment]
@pi-mst If there is interest, I have exactly the the sample code requested: WS2812B run from DMA and PIO, which is running my Christmas light display.  @nickovs If you want a block of the code to paste into your docs as an example, I can attach a simplified skeleton.

**Comment by nickovs** [comment]
I guess the question is how much code do we want to include in the docs themselves? Most of the reference pages do not include a great deal of example code.

I've tried to keep the three examples here short enough that you can see what they are doing, without having a ton of code that is doing something other than the DMA functionality. As it stands there are more examples in here than in the page for `rp2.PIO`, which is rather more complex.

Most more complex examples will end up more being an example of, say, programming the PIO, rather than using DMA. I'd argue now that DMA is merged, it might be better to start using DMA in the examples elsewhere, rather than having examples for use of PIO, SPI and other devices in the DMA docs.

**Comment by mendenm** [comment]
@nickovs The question about how much code is entirely valid.  The other possibility, given the unusual nature of the PIO/ and DMA devices on the RP2, is to create  a special examples section for the rp2 port.  Maybe each port should have an examples section that show off specific features of that port.  That  way, the examples are maintained as good code, and the documentation would link to the appropriate examples section.

File: docs/library/rp2.DMA.rst [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    Pack the values provided in the keyword arguments into the named fields of a new control
+    register value. Any field that is not provided will be set to a default value. The
+    default will either be taken from the provided ``default`` value, or if that is not
+    given, a default suitable for the current channel; setting this to the current value
+    of the `DMA.ctrl` attribute provides an easy way to override a subset of the fields.
+
+    The keys for the keyword arguments can be any key returned by the :meth:`DMA.unpack_ctrl()`
+    method. The writable values are:
+
+    :param enable: ``BOOL`` Set to enable the channel (default: ``True``).
```
Comment (projectgus) [inline docs/library/rp2.DMA.rst:120]: Nitpick: Suggest `bool` or `boolean` for these, to be inline with other docs and/or the Python type.

**Review by projectgus — APPROVED** [review]
This is really useful documentation, thanks for contributing it @nickovs!

**Comment by projectgus** [comment]
>  The other possibility, given the unusual nature of the PIO and DMA devices on the RP2, is to create a special examples section for the rp2 port

There is already an examples `examples/rp2` directory with some examples for the rp2 port (using PIO), so I think this would be the best outcome.

I think it would be OK to merge this now and add a more detailed DMA example later, though (maybe at the same time can remove the longer example from the .rst) - depending on Nick's schedule.

File: docs/library/rp2.DMA.rst (hunk 1 above)
Comment (dpgeorge) [reply docs/library/rp2.DMA.rst]: I have made this change during the rebase/merge.

**Comment by dpgeorge** [comment]
Now merged.  Thanks @nickovs for writing this up, it's always hard doing the docs.

**Pull request merged**